statev2/state/src/applicator/order_book.rs

```diff
@@ -0,0 +1,488 @@
+            .map_err(StateApplicatorError::Storage)?;
+        Self::attach_validity_proof_with_tx(&id, bundle, &tx)?;
+        let order_info = Self::read_order_info(&id, &tx)?;
+        tx.commit().map_err(StateApplicatorError::Storage)?;
+
+        self.system_bus().publish(
+            ORDER_STATE_CHANGE_TOPIC.to_string(),
+            SystemBusMessage::OrderStateChange { order: order_info },
+        );
+        Ok(())
+    }
+
+    /// Nullify orders indexed by a given wallet share nullifier
+    pub fn nullify_orders(&self, msg: NullifyOrdersMsg) -> Result<()> {
+        let nullifier: Scalar = msg.nullifier.unwrap_or_default().into();
+        let tx = self
+            .db()
+            .new_write_tx()
+            .map_err(StateApplicatorError::Storage)?;
+
+        self.nullify_orders_with_tx(nullifier, &tx)?;
+        tx.commit().map_err(StateApplicatorError::Storage)
+    }
+
+    // ------------------------
+    // | Order Update Helpers |
+    // ------------------------
+
+    /// Add an order to the book
+    ///
+    /// TODO: For an initial implementation we do not re-index based on local orders or
+    /// verified orders. This will be added with the getter implementations
+    fn add_order_with_tx(order: &NetworkOrder, tx: &DbTxn<'_, RW>) -> Result<()> {
+        // Write the order to storage
+        Self::write_order_info(order, tx)?;
+        // Add the order to the set of orders indexed by the nullifier
+        Self::append_to_nullifier_set(order.public_share_nullifier, order.id, tx)
+    }
+
+    /// Update the validity proof for an order
+    fn attach_validity_proof_with_tx(
+        order_id: &OrderIdentifier,
+        proof: OrderValidityProofBundle,
+        tx: &DbTxn<'_, RW>,
+    ) -> Result<()> {
+        // Read the order's current info
+        let mut order_info = Self::read_order_info(order_id, tx)?;
+
+        // Re-index based on the proof's nullifier
+        let prev_nullifier = order_info.public_share_nullifier;
+        let new_nullifier = proof.reblind_proof.statement.original_shares_nullifier;
+        if prev_nullifier != new_nullifier {
+            Self::update_order_nullifier(order_id, prev_nullifier, new_nullifier, tx)?;
+        }
+
+        // Update the order's info
+        order_info.state = NetworkOrderState::Verified;
+        order_info.public_share_nullifier = proof.reblind_proof.statement.original_shares_nullifier;
+        order_info.validity_proofs = Some(proof);
+        Self::write_order_info(&order_info, tx)
+    }
+
+    /// Cancel an order
+    fn cancel_order_with_tx(&self, order_id: &OrderIdentifier, tx: &DbTxn<'_, RW>) -> Result<()> {
+        let mut order = Self::read_order_info(order_id, tx)?;
+        order.state = NetworkOrderState::Cancelled;
+        order.validity_proof_witnesses = None;
+        order.validity_proofs = None;
+
+        Self::write_order_info(&order, tx)?;
+        self.system_bus().publish(
+            ORDER_STATE_CHANGE_TOPIC.to_string(),
```

Comment:
I think this is fine for now, but moving forward I'm a bit wary of properly tracking where/when we publish to the system bus. What I would want to avoid is flooding the system bus with messages redundant messages when multiple helpers get called / some helper gets called repeatedly, e.g. if we do multiple different updates on an order in a single state transition and publish to the system bus for each update. Would be nice to publish the minimum set of necessary messages to the system bus in one fell swoop at the end of a state transition / public function on this interface, so we can be sure that no redundant publishing is being done elsewhere in the helpers.

(to be clear, I understand that there are currently no redundant messages being publish to the system bus, I just want to be careful about this moving forward)

Reply:
Sure, any ideas on how to guarantee this?